cai_causal_graph/utils.py

```diff
@@ -0,0 +1,103 @@
+"""
+Copyright 2023 Impulse Innovations Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import re
+from typing import Dict, List, Tuple
+
+from cai_causal_graph.type_definitions import HasIdentifier, NodeLike
+
+
+def get_variable_name_and_lag(node_name: NodeLike) -> Tuple[str, int]:
+    """
+    Extract the variable name from a node name and time-series lag from a node name.
+
+    Example:
+        'X lag(n=2)' -> 'X', -2 if lagged in the past,
+        'X future(n=2)' -> 'X', 2 if lagged in the future.
+    """
+    # get the string name if the node is a Node object
+    if isinstance(node_name, HasIdentifier):
+        node_name = node_name.identifier
+
+    if not isinstance(node_name, str):
+        raise TypeError(f'Expected node name to be a string, got: {node_name}')
+
+    is_match = re.match(r'(\w+)(?: lag\(n=(\d+)\))?(?: future\(n=(\d+)\))?', node_name)
+
+    if is_match:
+        variable_name = is_match.group(1)
+        past_lag = is_match.group(2)
+        future_lag = is_match.group(3)
+
+        if past_lag:
+            return variable_name, -int(past_lag)
+        elif future_lag:
+            return variable_name, int(future_lag)
+        else:
+            return variable_name, 0  # no lag information
+
+    else:
+        raise ValueError(f'Invalid node name: {node_name}. Expected format: "X", "X lag(n=2)", or "X future(n=2)".')
+
+
+def get_name_from_lag(variable_name: str, lag: int) -> str:
+    """
+    Get the name of a lagged variable.
+
+    If the lag is 0, then the variable name is returned.
+    If the lag is not 0, then the old lag is removed and
+    variable name is appended with the lag.
+
+    Example:
+        'X', -2 -> 'X lag(n=2)', # lagged in the past
+        'X', 2 -> 'X future(n=2)', # lagged in the future
+
+    :param variable_name: The name of the variable.
+    :param lag: The lag of the variable.
+    :return: The name of the lagged variable.
+    """
+    # remove the old lag if it exists
+    variable_name, old_lag = get_variable_name_and_lag(variable_name)
```

Comment:
Nope because this function has the objective of getting the new lag even if there was one. I think it is confusing because I called it `variable_name`. I will change it.